So, if I try to work this into my code I get two errors:

```rust
error[E0599]: no method named `lex` found for type `lrlex::LexerDef<u16>` in the current scope
   --> src/lib/ast.rs:768:35
    |
768 |     let mut lexemes = match lexer.lex(&input) {
    |                                   ^^^

error: method `new` is private
   --> src/lib/ast.rs:772:18
    |
772 |     lexemes.push(Lexeme::new(u16::try_from(usize::from(grm.end_term_idx())).unwrap(),
    |                  ^^^^^^^^^^^

error: aborting due to previous error(s)
```

--- a/src/lib/lexer.rs
+++ b/src/lib/lexer.rs
@@ -30,8 +30,10 @@
 // DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 
+use std::cell::RefCell;
 use std::collections::HashMap;
 use std::convert::TryFrom;
+use std::rc::Rc;
 use std::slice::Iter;
 
 use regex::Regex;
@@ -106,12 +108,13 @@ impl<TokId: Copy + Eq> LexerDef<TokId> {
 /// tied to a single `String`, no guarantees are made about whether the lexemes are cached or not.
 pub struct Lexer<'a, TokId: 'a> {
     lexerdef: &'a LexerDef<TokId>,
-    s: &'a String
+    s: &'a String,
+    newlines: Rc<RefCell<Vec<usize>>>
 }
 
 impl<'a, TokId: Copy + Eq> Lexer<'a, TokId> {
     fn new(lexerdef: &'a LexerDef<TokId>, s: &'a String) -> Lexer<'a, TokId> {
-        Lexer {lexerdef, s}
+        Lexer {lexerdef, s, newlines: Rc::new(RefCell::new(Vec::new()))}
     }
 
     /// Return all this lexer's lexemes or a `LexError` if there was a problem when lexing.
@@ -134,6 +137,11 @@ impl<'a, TokId: Copy + Eq> Lexer<'a, TokId> {
                 }
             }
             if longest > 0 {
+                self.newlines.borrow_mut().extend(self.s[i .. i + longest]
+                                                      .chars()
+                                                      .enumerate()
+                                                      .filter(|&(_, c)| c == '\n')
+                                                      .map(|(j, _)| i + j + 1));
                 let r = &self.lexerdef.get_rule(longest_ridx).unwrap();
                 if r.name.is_some() {
                     lxs.push(Lexeme::new(r.tok_id, i, longest));
@@ -144,6 +152,22 @@ impl<'a, TokId: Copy + Eq> Lexer<'a, TokId> {
             }
         }
         Ok(lxs)
+    }
+
+    /// Return the line and column number of a `Lexeme`, or `Err` if it is clearly out of bounds
+    /// for this lexer.
+    pub fn line_and_col(&self, l: &Lexeme<TokId>) -> Result<(usize, usize), ()> {
+        if l.start > self.s.len() {
+            return Err(());
+        }
+        for (i, n) in self.newlines.borrow().iter().enumerate() {
+            if *n > l.start {
+                return Ok((i + 1, *n - l.start));
+            }
+        }
+        let newlines_brw = self.newlines.borrow();
+        let newlines_len = newlines_brw.len();
+        return Ok((newlines_len + 1, l.start - newlines_brw[newlines_len - 1] + 1))
     }
 }
 
@@ -242,7 +266,6 @@ if IF
         lexer.set_rule_ids(&map);
 
         let lexemes = lexer.lexer(&"iff if".to_string()).lexemes().unwrap();
-        println!("{:?}", lexemes);
         assert_eq!(lexemes.len(), 2);
         let lex1 = lexemes[0];
         assert_eq!(lex1.tok_id, 1);
@@ -252,5 +275,28 @@ if IF
         assert_eq!(lex2.tok_id, 0);
         assert_eq!(lex2.start, 4);
         assert_eq!(lex2.len, 2);
+    }
+
+    #[test]
+    fn test_line_and_col() {
+        let src = "%%
+[a-z]+ ID
+[ \\n] ;".to_string();
+        let mut lexer = parse_lex(&src).unwrap();
+        let mut map = HashMap::new();
+        map.insert("ID", 0);
+        lexer.set_rule_ids(&map);
+
+        let stream = " a\nb\n  c".to_string();
+        let lexer = lexer.lexer(&stream);
+        let lexemes = lexer.lexemes().unwrap();
+        assert_eq!(lexemes.len(), 3);
+        assert_eq!(lexer.line_and_col(&lexemes[0]).unwrap(), (1, 2));
+        assert_eq!(lexer.line_and_col(&lexemes[1]).unwrap(), (2, 2));
+        assert_eq!(lexer.line_and_col(&lexemes[2]).unwrap(), (3, 3));
+        let fake_lexeme = Lexeme{start: 100, len: 1, tok_id: 0};
+        if let Ok(_) = lexer.line_and_col(&fake_lexeme) {
+            panic!("line_and_col returned Ok(_) when it should have returned Err.");
+        }
     }
 }

--- a/src/lib/lexer.rs
+++ b/src/lib/lexer.rs
@@ -99,21 +99,21 @@ impl<TokId: Copy + Eq> LexerDef<TokId> {
     }
 
     /// Return a lexer for the `String` `s` that will lex relative to this `LexerDef`.
-    pub fn lexer<'a>(&'a self, s: &'a String) -> Lexer<'a, TokId> {
+    pub fn lexer<'a>(&'a self, s: &'a str) -> Lexer<'a, TokId> {
         Lexer::new(&self, s)
     }
 }
 
-/// A lexer holds a reference to a `String` and can lex it into `Lexeme`s. Although the struct is
-/// tied to a single `String`, no guarantees are made about whether the lexemes are cached or not.
+/// A lexer holds a reference to a string and can lex it into `Lexeme`s. Although the struct is
+/// tied to a single string, no guarantees are made about whether the lexemes are cached or not.
 pub struct Lexer<'a, TokId: 'a> {
     lexerdef: &'a LexerDef<TokId>,
-    s: &'a String,
+    s: &'a str,
     newlines: Rc<RefCell<Vec<usize>>>
 }
 
 impl<'a, TokId: Copy + Eq> Lexer<'a, TokId> {
-    fn new(lexerdef: &'a LexerDef<TokId>, s: &'a String) -> Lexer<'a, TokId> {
+    fn new(lexerdef: &'a LexerDef<TokId>, s: &'a str) -> Lexer<'a, TokId> {
         Lexer {lexerdef, s, newlines: Rc::new(RefCell::new(Vec::new()))}
     }
 
@@ -205,8 +205,6 @@ impl<TokId: Copy> Lexeme<TokId> {
     }
 }
 
-
-
 #[cfg(test)]
 mod test {
     use std::collections::HashMap;
@@ -227,7 +225,7 @@ mod test {
         map.insert("id", 1);
         lexer.set_rule_ids(&map);
 
-        let lexemes = lexer.lexer(&"abc 123".to_string()).lexemes().unwrap();
+        let lexemes = lexer.lexer(&"abc 123").lexemes().unwrap();
         assert_eq!(lexemes.len(), 2);
         let lex1 = lexemes[0];
         assert_eq!(lex1.tok_id, 1);
@@ -246,7 +244,7 @@ mod test {
 [0-9]+ int
         ".to_string();
         let lexer = parse_lex::<u8>(&src).unwrap();
-        match lexer.lexer(&"abc".to_string()).lexemes() {
+        match lexer.lexer(&"abc").lexemes() {
             Ok(_)  => panic!("Invalid input lexed"),
             Err(LexError{idx: 0}) => (),
             Err(e) => panic!("Incorrect error returned {:?}", e)
@@ -265,7 +263,7 @@ if IF
         map.insert("ID", 1);
         lexer.set_rule_ids(&map);
 
-        let lexemes = lexer.lexer(&"iff if".to_string()).lexemes().unwrap();
+        let lexemes = lexer.lexer(&"iff if").lexemes().unwrap();
         assert_eq!(lexemes.len(), 2);
         let lex1 = lexemes[0];
         assert_eq!(lex1.tok_id, 1);
@@ -287,7 +285,7 @@ if IF
         map.insert("ID", 0);
         lexer.set_rule_ids(&map);
 
-        let stream = " a\nb\n  c".to_string();
+        let stream = " a\nb\n  c";
         let lexer = lexer.lexer(&stream);
         let lexemes = lexer.lexemes().unwrap();
         assert_eq!(lexemes.len(), 3);

--- a/src/lib/lexer.rs
+++ b/src/lib/lexer.rs
@@ -52,13 +52,15 @@ pub struct LexError {
     idx: usize
 }
 
-pub struct Lexer<TokId> {
+/// This struct represents, in essence, a .l file in memory. From it one can produce a `Lexer`
+/// which actually lexes inputs.
+pub struct LexerDef<TokId> {
     rules: Vec<Rule<TokId>>
 }
 
-impl<TokId: Copy + Eq> Lexer<TokId> {
-    pub fn new(rules: Vec<Rule<TokId>>) -> Lexer<TokId> {
-        Lexer {rules: rules}
+impl<TokId: Copy + Eq> LexerDef<TokId> {
+    pub fn new(rules: Vec<Rule<TokId>>) -> LexerDef<TokId> {
+        LexerDef{rules: rules}
     }
 
     /// Get the `Rule` at index `idx`.
@@ -94,15 +96,34 @@ impl<TokId: Copy + Eq> Lexer<TokId> {
         self.rules.iter()
     }
 
-    pub fn lex(&self, s: &str) -> Result<Vec<Lexeme<TokId>>, LexError> {
+    /// Return a lexer for the `String` `s` that will lex relative to this `LexerDef`.
+    pub fn lexer<'a>(&'a self, s: &'a String) -> Lexer<'a, TokId> {
+        Lexer::new(&self, s)
+    }
+}
+
+/// A lexer holds a reference to a `String` and can lex it into `Lexeme`s. Although the struct is
+/// tied to a single `String`, no guarantees are made about whether the lexemes are cached or not.
+pub struct Lexer<'a, TokId: 'a> {
+    lexerdef: &'a LexerDef<TokId>,
+    s: &'a String
+}
+
+impl<'a, TokId: Copy + Eq> Lexer<'a, TokId> {
+    fn new(lexerdef: &'a LexerDef<TokId>, s: &'a String) -> Lexer<'a, TokId> {
+        Lexer {lexerdef, s}
+    }
+
+    /// Return all this lexer's lexemes or a `LexError` if there was a problem when lexing.
+    pub fn lexemes(&self) -> Result<Vec<Lexeme<TokId>>, LexError> {
         let mut i = 0; // byte index into s
         let mut lxs = Vec::new(); // lexemes
 
-        while i < s.len() {
+        while i < self.s.len() {
             let mut longest = 0; // Length of the longest match
             let mut longest_ridx = 0; // This is only valid iff longest != 0
-            for (ridx, r) in self.iter_rules().enumerate() {
-                if let Some(m) = r.re.find(&s[i..]) {
+            for (ridx, r) in self.lexerdef.iter_rules().enumerate() {
+                if let Some(m) = r.re.find(&self.s[i..]) {
                     let len = m.end();
                     // Note that by using ">", we implicitly prefer an earlier over a later rule, if
                     // both match an input of the same length.
@@ -113,7 +134,7 @@ impl<TokId: Copy + Eq> Lexer<TokId> {
                 }
             }
             if longest > 0 {
-                let r = &self.get_rule(longest_ridx).unwrap();
+                let r = &self.lexerdef.get_rule(longest_ridx).unwrap();
                 if r.name.is_some() {
                     lxs.push(Lexeme::new(r.tok_id, i, longest));
                 }
@@ -126,6 +147,8 @@ impl<TokId: Copy + Eq> Lexer<TokId> {
     }
 }
 
+/// A lexeme has a starting position in a string, a length, and a token id. It is a deliberately
+/// small data-structure to large input files to be stored efficiently.
 #[derive(Clone, Copy, Debug)]
 pub struct Lexeme<TokId> {
     start: usize,
@@ -158,6 +181,8 @@ impl<TokId: Copy> Lexeme<TokId> {
     }
 }
 
+
+
 #[cfg(test)]
 mod test {
     use std::collections::HashMap;
@@ -178,7 +203,7 @@ mod test {
         map.insert("id", 1);
         lexer.set_rule_ids(&map);
 
-        let lexemes = lexer.lex("abc 123").unwrap();
+        let lexemes = lexer.lexer(&"abc 123".to_string()).lexemes().unwrap();
         assert_eq!(lexemes.len(), 2);
         let lex1 = lexemes[0];
         assert_eq!(lex1.tok_id, 1);
@@ -197,7 +222,7 @@ mod test {
 [0-9]+ int
         ".to_string();
         let lexer = parse_lex::<u8>(&src).unwrap();
-        match lexer.lex("abc") {
+        match lexer.lexer(&"abc".to_string()).lexemes() {
             Ok(_)  => panic!("Invalid input lexed"),
             Err(LexError{idx: 0}) => (),
             Err(e) => panic!("Incorrect error returned {:?}", e)
@@ -216,7 +241,7 @@ if IF
         map.insert("ID", 1);
         lexer.set_rule_ids(&map);
 
-        let lexemes = lexer.lex("iff if").unwrap();
+        let lexemes = lexer.lexer(&"iff if".to_string()).lexemes().unwrap();
         println!("{:?}", lexemes);
         assert_eq!(lexemes.len(), 2);
         let lex1 = lexemes[0];

--- a/src/lib/mod.rs
+++ b/src/lib/mod.rs
@@ -40,7 +40,7 @@ use std::fmt;
 mod lexer;
 mod parser;
 
-pub use lexer::{Lexeme, Lexer};
+pub use lexer::{Lexeme, LexerDef, Lexer};
 use parser::parse_lex;
 
 #[macro_use]
@@ -84,7 +84,7 @@ impl fmt::Display for LexBuildError {
     }
 }
 
-pub fn build_lex<TokId: Copy + Eq + TryFrom<usize>>(s: &str) -> Result<Lexer<TokId>, LexBuildError> {
+pub fn build_lex<TokId: Copy + Eq + TryFrom<usize>>(s: &str) -> Result<LexerDef<TokId>, LexBuildError> {
     parse_lex(s)
 }
 

--- a/src/lib/parser.rs
+++ b/src/lib/parser.rs
@@ -35,7 +35,7 @@ use std::convert::TryFrom;
 use regex::{Regex, RegexBuilder};
 
 use {LexErrorKind, LexBuildError, LexBuildResult};
-use lexer::{Lexer, Rule};
+use lexer::{LexerDef, Rule};
 
 pub struct LexParser<TokId> {
     src: String,
@@ -187,8 +187,8 @@ impl<TokId: TryFrom<usize>> LexParser<TokId> {
     }
 }
 
-pub fn parse_lex<TokId: Copy + Eq + TryFrom<usize>>(s: &str) -> LexBuildResult<Lexer<TokId>> {
-    LexParser::new(s.to_string()).map(|p| Lexer::new(p.rules))
+pub fn parse_lex<TokId: Copy + Eq + TryFrom<usize>>(s: &str) -> LexBuildResult<LexerDef<TokId>> {
+    LexParser::new(s.to_string()).map(|p| LexerDef::new(p.rules))
 }
 
 #[cfg(test)]

--- a/src/main.rs
+++ b/src/main.rs
@@ -39,7 +39,7 @@ use std::fs::File;
 use std::io::{Read, stderr, Write};
 use std::path::Path;
 
-use lrlex::{build_lex, Lexer};
+use lrlex::{build_lex, LexerDef};
 
 fn usage(prog: String, msg: &str) {
     let path = Path::new(prog.as_str());
@@ -85,13 +85,13 @@ fn main() {
     }
 
     let lex_l_path = &matches.free[0];
-    let lexer: Lexer<usize> = build_lex(&read_file(lex_l_path)).unwrap_or_else(|s| {
+    let lexerdef: LexerDef<usize> = build_lex(&read_file(lex_l_path)).unwrap_or_else(|s| {
             writeln!(&mut stderr(), "{}: {}", &lex_l_path, &s).ok();
             process::exit(1);
         });
     let input = &read_file(&matches.free[1]);
-    let lexemes = lexer.lex(&input).unwrap();
+    let lexemes = lexerdef.lexer(&input).lexemes().unwrap();
     for l in lexemes.iter() {
-        println!("{} {}", lexer.get_rule_by_id(l.tok_id()).unwrap().name.as_ref().unwrap(), &input[l.start()..l.start() + l.len()]);
+        println!("{} {}", lexerdef.get_rule_by_id(l.tok_id()).unwrap().name.as_ref().unwrap(), &input[l.start()..l.start() + l.len()]);
     }
 }

